speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backend/SQLDatabaseStorage.cpp

```diff
@@ -1,229 +1,487 @@
-/// @param p_variableAmount how many variables the table has
-/// @param p_headers        header names
-void CreateTableSQL(
-        std::ifstream& p_inputFile,
-        std::string& p_reading,
-        const std::string& p_tableName,
-        std::stringstream& p_sqlStream,
-        const std::unordered_map<std::string,std::string>& p_headerTypes,
-        int& p_variableAmount,
-        std::string& p_headers)
+/// @brief Connect to the specified database. Initialise the database with the proper tables if they don't exist yet.
+/// @param p_hostName Hostname of the database to connect to. Should be a TCP-IP address.
+/// @param p_port Port the database is located on on the host.
+/// @param p_username Username to connect with to the database.
+/// @param p_password Password to connect with to the database.
+/// @param p_schemaName Name of the database schema to use.
+/// @return returns true if connection to database has been made, false otherwise
+bool SQLDatabaseStorage::OpenDatabase(
+    const std::string& p_hostName,
+    int p_port,
+    const std::string& p_username,
+    const std::string& p_password,
+    const std::string& p_schemaName)
 {
-    std::stringstream headersStream;
+    // Initialise SQL driver
+    m_driver = sql::mysql::get_mysql_driver_instance();
 
-    p_sqlStream.clear();
-    p_sqlStream.str("");
+    // Set connection options, and connect to the database
+    sql::ConnectOptionsMap connection_properties;
+    connection_properties["hostName"] = "tcp://" + p_hostName;
+    connection_properties["userName"] = p_username;
+    connection_properties["password"] = p_password;
+    connection_properties["port"] = p_port;
+    connection_properties["OPT_RECONNECT"] = true;
+    connection_properties["CLIENT_MULTI_STATEMENTS"] = false;
+    try { m_connection = m_driver->connect(connection_properties);}
+    catch (std::exception& e) {return false;}
 
-    p_sqlStream << "CREATE TABLE " << p_tableName << " (TimeStamp int NOT NULL PRIMARY KEY, ";
+    // Create the database schema if this is a new schema. This has to be done before setting the schema on the connection.
+    m_statement = m_connection->createStatement();
 
-    p_inputFile >> p_reading;
+    EXECUTE("CREATE DATABASE IF NOT EXISTS " + p_schemaName)
+    m_statement->close();
+    delete m_statement;
 
-    if (p_reading == "START") throw std::exception("Variable names were not given");
+    // Set the correct database schema
+    m_connection->setSchema(p_schemaName);
 
-    // starts at one because of timestamp
-    int variableAmount = 1;
+    // Create a (reusable) statement
+    m_statement = m_connection->createStatement();
 
-    while (true)
-    {
-        variableAmount++;
-        headersStream << p_reading;
-        try
-        {
-            p_sqlStream << p_headerTypes.at(p_reading);
-        }
-        catch (const std::exception& e) {throw std::exception("Variable does not exist in map");}
```

Comment:
throwing an exception when you can also check if the key exists in the map

Reply:
This is removed code?